jsonrpc/codec.go

```diff
@@ -102,6 +103,77 @@ const (
 
 type BlockNumber int64
 
+type BlockNumberOrHash struct {
+	BlockNumber *BlockNumber `json:"blockNumber,omitempty"`
+	BlockHash   *types.Hash  `json:"blockHash,omitempty"`
+}
+
+// Unmarshal will try to extract the filter's data.
+// Here are the possible input formats :
+//
+// 1 - "latest", "pending" or "earliest"	- self-explaining keywords
+// 2 - "0x2"								- block number #2 (EIP-1898 backward compatible)
+// 3 - {blockNumber:	"0x2"}				- EIP-1898 compliant block number #2
+// 4 - {blockHash:		"0xe0e..."}			- EIP-1898 compliant block hash 0xe0e...
+func (bnh *BlockNumberOrHash) Unmarshal(input *interface{}) error {
+	var placeholder BlockNumberOrHash
+
+	data, err := json.Marshal(*input)
+	if err != nil {
+		return fmt.Errorf("failed to serialize input: %v", err)
+	}
+
+	err = json.Unmarshal(data, &placeholder)
+	if err != nil {
+		var keyword string
+		err = json.Unmarshal(data, &keyword)
+		if err == nil {
+			// Try to extract keyword
+			switch keyword {
+			case "pending":
+				n := PendingBlockNumber
+				bnh.BlockNumber = &n
+				return nil
+			case "latest":
+				n := LatestBlockNumber
+				bnh.BlockNumber = &n
+				return nil
+			case "earliest":
+				n := EarliestBlockNumber
+				bnh.BlockNumber = &n
+				return nil
+			default:
+				// Try to extract hex number
+				s, ok := (*input).(string)
+				if !ok {
+					return fmt.Errorf("input cannot be converted to string")
+				}
+				if len(s) < 3 || !strings.HasPrefix(s, "0x") {
+					return fmt.Errorf("invalid hexadecimal number provided for block number")
+				}
+				number, err := strconv.ParseInt(s[2:], 16, 64)
+				fmt.Println(number, err)
+				if err != nil {
+					return fmt.Errorf("failed to convert hex string to int64: %v", err)
+				}
+				bnh.BlockNumber = (*BlockNumber)(&number)
+				return nil
+			}
+		}
+		return fmt.Errorf("invalid block number provided")
+	}
+
+	// Try to extract object
+	bnh.BlockNumber = placeholder.BlockNumber
+	bnh.BlockHash = placeholder.BlockHash
+
+	if bnh.BlockNumber != nil && bnh.BlockHash != nil {
+		return fmt.Errorf("cannot get account balance using both block number hash")
+	}
+
```

Comment:
There is a possibility that both `bnh.BlockNumber` and `bnh.BlockHash` is `nil`. In this case, panic occurs when getting its value. Actually bnh.BlockNumber which may be nill is read in `GetTransactionCount`. 

```go
if bnh.BlockHash != nil {
                 // safe to access to bnh.BlockHash
		block, ok := e.d.store.GetBlockByHash(*bnh.BlockHash, false)
		if !ok {
			return nil, fmt.Errorf("could not find block referenced by the hash %s", bnh.BlockHash.String())
		}
                ...
	} else {
                 // unsafe to access to bnh.BlockNumber!
		nonce, err = e.d.getNextNonce(address, *bnh.BlockNumber)
	}
```

By the way, such values can be got by passing unexpected value like:

```json
{
	"jsonrpc":"2.0",
	"method":"eth_getTransactionCount",
	"params":[
		"0xBa8CAE91c858120E9699f9b9936418C6bd3ecCC1",
        {"hoge": "0x1"}
	],
	"id":1
}
```

So my suggestion is that we need to return error when both BlockHash and BlockNumber is nil